src/init.c

```diff
@@ -209,8 +213,12 @@ shmem_internal_init(int tl_requested, int *tl_provided)
     shmem_internal_data_base = (void*) get_etext();
     shmem_internal_data_length = get_end() - get_etext();
 #else
-    shmem_internal_data_base = &data_start;
-    shmem_internal_data_length = (unsigned long) &end  - (unsigned long) &data_start;
+    if (&__data_start == 0 || &_end == 0)
```

Comment:
Should probably cast 0 to (int*).

Also, need a comment to record that if the weak symbol isn't bound, its address will be 0.